// @defichain/jellyfish-transaction-builder/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.P2WPKHTransactionBuilder = void 0;
const txn_builder_1 = require("./txn/txn_builder");
const txn_builder_dex_1 = require("./txn/txn_builder_dex");
const txn_builder_utxo_1 = require("./txn/txn_builder_utxo");
const txn_builder_account_1 = require("./txn/txn_builder_account");
const txn_builder_oracles_1 = require("./txn/txn_builder_oracles");
const txn_builder_liq_pool_1 = require("./txn/txn_builder_liq_pool");
const txn_builder_governance_1 = require("./txn/txn_builder_governance");
const txn_builder_icxorderbook_1 = require("./txn/txn_builder_icxorderbook");
const txn_builder_masternode_1 = require("./txn/txn_builder_masternode");
const txn_builder_loans_1 = require("./txn/txn_builder_loans");
const txn_builder_vault_1 = require("./txn/txn_builder_vault");
__exportStar(require("./provider"), exports);
__exportStar(require("./txn/txn_fee"), exports);
__exportStar(require("./txn/txn_builder"), exports);
__exportStar(require("./txn/txn_builder_dex"), exports);
__exportStar(require("./txn/txn_builder_utxo"), exports);
__exportStar(require("./txn/txn_builder_account"), exports);
__exportStar(require("./txn/txn_builder_oracles"), exports);
__exportStar(require("./txn/txn_builder_loans"), exports);
__exportStar(require("./txn/txn_builder_vault"), exports);
__exportStar(require("./txn/txn_builder_liq_pool"), exports);
__exportStar(require("./txn/txn_builder_icxorderbook"), exports);
__exportStar(require("./txn/txn_builder_masternode"), exports);
/**
 * All in one transaction builder.
 * Currently only support sending from P2PKH operations.
 */
class P2WPKHTransactionBuilder extends txn_builder_1.P2WPKHTxnBuilder {
    constructor() {
        super(...arguments);
        this.dex = new txn_builder_dex_1.TxnBuilderDex(this.feeProvider, this.prevoutProvider, this.ellipticPairProvider, this.network);
        this.utxo = new txn_builder_utxo_1.TxnBuilderUtxo(this.feeProvider, this.prevoutProvider, this.ellipticPairProvider, this.network);
        this.account = new txn_builder_account_1.TxnBuilderAccount(this.feeProvider, this.prevoutProvider, this.ellipticPairProvider, this.network);
        this.liqPool = new txn_builder_liq_pool_1.TxnBuilderLiqPool(this.feeProvider, this.prevoutProvider, this.ellipticPairProvider, this.network);
        this.oracles = new txn_builder_oracles_1.TxnBuilderOracles(this.feeProvider, this.prevoutProvider, this.ellipticPairProvider, this.network);
        this.governance = new txn_builder_governance_1.TxnBuilderGovernance(this.feeProvider, this.prevoutProvider, this.ellipticPairProvider, this.network);
        this.icxorderbook = new txn_builder_icxorderbook_1.TxnBuilderICXOrderBook(this.feeProvider, this.prevoutProvider, this.ellipticPairProvider, this.network);
        this.masternode = new txn_builder_masternode_1.TxnBuilderMasternode(this.feeProvider, this.prevoutProvider, this.ellipticPairProvider, this.network);
        this.loans = new txn_builder_loans_1.TxnBuilderLoans(this.feeProvider, this.prevoutProvider, this.ellipticPairProvider, this.network);
        this.vault = new txn_builder_vault_1.TxnBuilderVault(this.feeProvider, this.prevoutProvider, this.ellipticPairProvider, this.network);
    }
}
exports.P2WPKHTransactionBuilder = P2WPKHTransactionBuilder;
