// @defichain/jellyfish-transaction/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fromOPCodes = exports.toOPCodes = exports.DeFiTransactionConstants = void 0;
/**
 * DeFi Blockchain Transaction Constants
 * https://github.com/DeFiCh/ain/blob/master/src/primitives/transaction.h
 */
exports.DeFiTransactionConstants = {
    Version: 0x00000004,
    WitnessMarker: 0x00,
    WitnessFlag: 0x01 // 1 byte
};
__exportStar(require("./script"), exports);
__exportStar(require("./tx"), exports);
__exportStar(require("./tx_composer"), exports);
__exportStar(require("./tx_segwit"), exports);
/**
 * Need to refactor this interface, it's not consistent
 */
var _buffer_1 = require("./script/_buffer");
Object.defineProperty(exports, "toOPCodes", { enumerable: true, get: function () { return _buffer_1.toOPCodes; } });
Object.defineProperty(exports, "fromOPCodes", { enumerable: true, get: function () { return _buffer_1.toBuffer; } });
