// @defichain/jellyfish-transaction/dist/script/expansion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OP_NOP10 = exports.OP_NOP9 = exports.OP_NOP8 = exports.OP_NOP7 = exports.OP_NOP6 = exports.OP_NOP5 = exports.OP_NOP4 = exports.OP_NOP3 = exports.OP_CHECKSEQUENCEVERIFY = exports.OP_NOP2 = exports.OP_CHECKLOCKTIMEVERIFY = exports.OP_NOP1 = void 0;
const opcode_1 = require("./opcode");
/**
 * This instruction performs no operation.
 * No items are added to the stack.
 */
class OP_NOP1 extends opcode_1.StaticCode {
    constructor() {
        super(0xb0, 'OP_NOP1');
    }
}
exports.OP_NOP1 = OP_NOP1;
/**
 * This instruction allows a transaction output to be made unspendable until some point in the future.
 */
class OP_CHECKLOCKTIMEVERIFY extends opcode_1.StaticCode {
    constructor() {
        super(0xb1, 'OP_CHECKLOCKTIMEVERIFY');
    }
}
exports.OP_CHECKLOCKTIMEVERIFY = OP_CHECKLOCKTIMEVERIFY;
/**
 * This instruction allows a transaction output to be made unspendable until some point in the future.
 * @See OP_CHECKLOCKTIMEVERIFY
 */
class OP_NOP2 extends OP_CHECKLOCKTIMEVERIFY {
}
exports.OP_NOP2 = OP_NOP2;
/**
 * This instruction allows execution pathways of a script to be restricted based on the age of the output being spent.
 */
class OP_CHECKSEQUENCEVERIFY extends opcode_1.StaticCode {
    constructor() {
        super(0xb2, 'OP_CHECKSEQUENCEVERIFY');
    }
}
exports.OP_CHECKSEQUENCEVERIFY = OP_CHECKSEQUENCEVERIFY;
/**
 * This instruction allows execution pathways of a script to be restricted based on the age of the output being spent.
 * @See OP_CHECKSEQUENCEVERIFY
 */
class OP_NOP3 extends OP_CHECKSEQUENCEVERIFY {
}
exports.OP_NOP3 = OP_NOP3;
/**
 * This instruction performs no operation.
 * No items are added to the stack.
 */
class OP_NOP4 extends opcode_1.StaticCode {
    constructor() {
        super(0xb3, 'OP_NOP4');
    }
}
exports.OP_NOP4 = OP_NOP4;
/**
 * This instruction performs no operation.
 * No items are added to the stack.
 */
class OP_NOP5 extends opcode_1.StaticCode {
    constructor() {
        super(0xb4, 'OP_NOP5');
    }
}
exports.OP_NOP5 = OP_NOP5;
/**
 * This instruction performs no operation.
 * No items are added to the stack.
 */
class OP_NOP6 extends opcode_1.StaticCode {
    constructor() {
        super(0xb5, 'OP_NOP6');
    }
}
exports.OP_NOP6 = OP_NOP6;
/**
 * This instruction performs no operation.
 * No items are added to the stack.
 */
class OP_NOP7 extends opcode_1.StaticCode {
    constructor() {
        super(0xb6, 'OP_NOP7');
    }
}
exports.OP_NOP7 = OP_NOP7;
/**
 * This instruction performs no operation.
 * No items are added to the stack.
 */
class OP_NOP8 extends opcode_1.StaticCode {
    constructor() {
        super(0xb7, 'OP_NOP8');
    }
}
exports.OP_NOP8 = OP_NOP8;
/**
 * This instruction performs no operation.
 * No items are added to the stack.
 */
class OP_NOP9 extends opcode_1.StaticCode {
    constructor() {
        super(0xb8, 'OP_NOP9');
    }
}
exports.OP_NOP9 = OP_NOP9;
/**
 * This instruction performs no operation.
 * No items are added to the stack.
 */
class OP_NOP10 extends opcode_1.StaticCode {
    constructor() {
        super(0xb9, 'OP_NOP10');
    }
}
exports.OP_NOP10 = OP_NOP10;
